File: SMores/NumA/Base Functions/solve_square_matrix.py
class Matrix:
    def __init__(self, matrix):
        self.matrix = matrix
        self.n = len(self.matrix)
        self.sol = []
    
    def forward_elimination(self):
        for k in range(self.n - 1):  # loop through steps
            for i in range(k + 1, len(self.matrix)):  # loop through rows in bottom left corner
                if self.matrix[k][k] == 0.0:
                    print("pivoting")
                    self.partial_pivoting(k, k)
                m = self.matrix[i][k] / self.matrix[k][k]
                for j in range(len(self.matrix) + 1):  # loop through columns
                    self.matrix[i][j] = self.matrix[i][j] - m * self.matrix[k][j]
            print(self)
    
    def backward_substitution(self):
        sol = [0] * self.n
        for row in range(self.n - 1, -1, -1):  # loop through rows backwards
            sum = self.matrix[row][self.n]  # value in right-most column
            for col in range(row + 1, self.n):  # loop through columns
                sum -= self.matrix[row][col] * sol[col]  # subtract each known value
            sol[row] = sum / self.matrix[row][row]  # divide by the diag value
        self.sol = sol
    
    def partial_pivoting(self, r, c):
        row = 0
        largest = 0
        # find the row with the largest value for the input column
        for i in range(self.n):
            if abs(self.matrix[i][c]) > largest:
                largest = self.matrix[i][c]
                row = i
        # swap the row found above with the input row
        temp = self.matrix[r]
        self.matrix[r] = self.matrix[row]
        self.matrix[row] = temp
    
    def scale_matrix(self):
        for i in range(len(self.matrix) - 1):
            largest = 0
            for j in range(len(self.matrix[i])):
                if abs(self.matrix[i][j]) > largest:
                    largest = self.matrix[i][j]
            for j in range(len(self.matrix[i])):
                self.matrix[i][j] = self.matrix[i][j] / largest
    
    def get_spot(self, i, j):
        return self.matrix[i][j]
    
    def solution(self):
        return '[' + ', '.join(['{}'.format(str(item)) for item in self.sol]) + ']\n'
    
    def __str__(self):
        mx = max((len(str(i)) for row in self.matrix for i in row))
        output = '['
        output += ']\n['.join([', '.join(['{:6.2f}'.format(i, mx = mx) for i in r]) for r in self.matrix])
        return output + ']\n'


n = 4
# x - y + 3 * z = -3
# -x - 2 * z = 1
# 2 * x + 2 * y + 4 * z = 0

# mat = Matrix([[1, 1, 0, 3, 4],
#               [2, 1, -1, 1, 1],
#               [3, -1, -1, 2, -3],
#               [-1, 2, 3, -1, 4]])
# mat = Matrix([[1, -1, 3, -3],
#               [-1, 0, -2, 1],
#               [2, 2, 4, 0]])
# mat = Matrix([[1, -5, 1, 7],
#               [10, 0, 20, 6],
#               [5, 0, -1, 4]])
mat = Matrix([[2, 0, 0, 0, 3],
              [1, 1.5, 0, 0, 4.5],
              [0, -3, 0.5, 0, -6.6],
              [2, -2, 1, 1, 0.8]])

print(mat, '\nForward Elimination:\n')
mat.forward_elimination()  # WORKS!!
print(mat, '\nBackward Substitution:\n')
mat.backward_substitution()
print(mat.solution())

File: SMores/NumA/Base Functions/solve_square_matrix.f90
program solve_matrix
    IMPLICIT NONE
    integer i, j, k, n
    doubleprecision m, sum
    doubleprecision a(4, 4), x(4)

    !    n = 4
    !    a(1, :) = [  1, 1, 0, 3, 4 ]
    !    a(2, :) = [  2, 1, -1, 1, 1 ]
    !    a(3, :) = [  3, -1, -1, 2, -3 ]
    !    a(4, :) = [ -1, 2, 3, -1, 4 ]
    !    a(5, :) = [  0, 0, 0, 0, 0 ]

    !(/ 1,  1,  0,  3,  4, 0, &
    !  2,  1, -1,  1,  1, 0, &
    ! 3, -1, -1,  2, -3, 0, &
    !-1,  2,  3, -1,  4, 0/)

    !Lab 4 Matrix:
    n = 4
    a(1, :) = [  1, -1, 3, -3 ]
    a(2, :) = [ -1, 0, -2, 1 ]
    a(3, :) = [  2, 2, 4, 0 ]
    a(4, :) = [  0, 0, 0, 0 ]

    write(*, *) print_matrix(a, n)
    write(*, *) ""
    !partial_pivot(r, c, a, n)
    !call partial_pivot(1, 1, 1, a, n)
    !write(*, *) print_matrix(a, n)
    stop

    !Forward Elimination
    do k = 1, n - 1
        do i = k + 1, n
            m = a(i, k) / a(k, k)
            do j = k + 1, n
                a(i, j) = a(i, j) - (m * a(k, j))
            ENDDO
        ENDDO
    ENDDO

    !Backward Substitution
    do i = n - 1, 1, -1
        sum = a(i, n + 1)
        do j = i + 1, n
            sum = sum - a(i, j) - x(j)
        end do
        x(i) = sum / a(i, i)
    ENDDO

    do i = 1, n
        write(*, '(a)') "x", i, "=", x(i)
    end do

CONTAINS

    SUBROUTINE partial_pivot(r, c, k, a, n)
        IMPLICIT NONE
        INTEGER :: n, k, r, c
        DOUBLEPRECISION, DIMENSION(n, n) :: a(n, n)
        integer row, col, i
        doubleprecision largest, temp

        !find row with largest column
        row = -1
        largest = 0
        DO i = k, n
            IF(a(i, c) > largest) THEN
                largest = abs(a(i, c))
                row = i
            END IF
        END DO

        !swap the two rows
        DO j = k, n
            temp = a(c, j)
            a(c, j) = a(k, j)
            a(k, j) = temp
        END DO

    END SUBROUTINE partial_pivot

    INTEGER FUNCTION print_matrix(a, n)
        IMPLICIT NONE
        INTEGER :: n
        DOUBLEPRECISION, DIMENSION(n, n) :: a(n, n)
        integer i, j
        character*100 output, num

        write(*, *) "Printing the matrix..."

        !        output = ""
        !        DO i = 1, n
        !
        !            output = output // "["
        !            DO j = 1, n
        !                write(num, "(A5,I2)") a(i, j)
        !                output = output // num // ", "
        !            END DO
        !            output = output // "]"
        !        END DO
        !        write(*, *) output
        !        print_matrix = 0

        do i = 1, n
            write(*, "(100g15.5)") (a(i, j), j = 1, n)
        enddo
    END FUNCTION print_matrix
END PROGRAM solve_matrix